File: disk_speed.cs
// disk_speed.cs — C# версия

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;

class DiskStats {
    public List<double> ReadSpeeds { get; set; } = new List<double>();
    public List<double> WriteSpeeds { get; set; } = new List<double>();
    public double MaxRead { get; set; } = 0;
    public double MaxWrite { get; set; } = 0;
}

class Program {
    private static int interval = 2;
    private static Dictionary<string, DiskStats> stats = new Dictionary<string, DiskStats>();
    private static Dictionary<string, long> prevRead = new Dictionary<string, long>();
    private static Dictionary<string, long> prevWrite = new Dictionary<string, long>();

    static void Main(string[] args) {
        if (args.Length > 0 && int.TryParse(args[0], out int val) && val > 0) {
            interval = val;
        }

        Console.WriteLine("\u001B[36m💾 Disk Speed Monitor (C#)\u001B[0m");
        Console.WriteLine($"Интервал: {interval} сек");
        Console.WriteLine("Нажмите Ctrl+C для остановки...");

        // Получаем начальную статистику
        GetDiskIO(out prevRead, out prevWrite);
        Thread.Sleep(interval * 1000);

        var timer = new Timer(_ => {
            try {
                GetDiskIO(out var currRead, out var currWrite);
                var speeds = CalculateSpeeds(prevRead, prevWrite, currRead, currWrite);
                PrintSpeeds(speeds);
                prevRead = currRead;
                prevWrite = currWrite;
            } catch (Exception e) {
                Console.WriteLine($"\u001B[31m❌ Ошибка: {e.Message}\u001B[0m");
            }
        }, null, 0, interval * 1000);

        Console.CancelKeyPress += (sender, e) => {
            e.Cancel = true;
            timer.Dispose();
            Console.WriteLine("\n\n⏹️ Остановка...");
            PrintStats();
            Environment.Exit(0);
        };

        Thread.Sleep(Timeout.Infinite);
    }

    static void GetDiskIO(out Dictionary<string, long> read, out Dictionary<string, long> write) {
        read = new Dictionary<string, long>();
        write = new Dictionary<string, long>();

        if (Environment.OSVersion.Platform == PlatformID.Win32NT) {
            // Windows: используем PerformanceCounter
            var drives = DriveInfo.GetDrives().Where(d => d.IsReady);
            foreach (var drive in drives) {
                read[drive.Name.Replace("\\", "")] = 0;
                write[drive.Name.Replace("\\", "")] = 0;
            }
        } else {
            // Linux/macOS: используем iostat (упрощённо)
            var psi = new ProcessStartInfo {
                FileName = "iostat",
                Arguments = "-d 1 2",
                RedirectStandardOutput = true,
                UseShellExecute = false
            };
            using var process = Process.Start(psi);
            string output = process.StandardOutput.ReadToEnd();
            process.WaitForExit();
            // Парсинг iostat (упрощённо)
            var lines = output.Split('\n');
            foreach (var line in lines) {
                var parts = line.Trim().Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length >= 6 && double.TryParse(parts[5], out _)) {
                    read[parts[0]] = (long)(double.Parse(parts[5]) * 1024);
                    write[parts[0]] = (long)(double.Parse(parts[6]) * 1024);
                }
            }
        }
    }

    static Dictionary<string, (double read, double write)> CalculateSpeeds(
        Dictionary<string, long> prevR, Dictionary<string, long> prevW,
        Dictionary<string, long> currR, Dictionary<string, long> currW) {

        var speeds = new Dictionary<string, (double, double)>();
        foreach (var disk in currR.Keys) {
            if (prevR.ContainsKey(disk) && currR.ContainsKey(disk)) {
                double readMB = (currR[disk] - prevR[disk]) / (1024.0 * 1024.0);
                double writeMB = (currW[disk] - prevW[disk]) / (1024.0 * 1024.0);
                speeds[disk] = (readMB / interval, writeMB / interval);

                if (!stats.ContainsKey(disk)) {
                    stats[disk] = new DiskStats();
                }
                stats[disk].ReadSpeeds.Add(readMB / interval);
                stats[disk].WriteSpeeds.Add(writeMB / interval);
                if (readMB / interval > stats[disk].MaxRead) stats[disk].MaxRead = readMB / interval;
                if (writeMB / interval > stats[disk].MaxWrite) stats[disk].MaxWrite = writeMB / interval;
            }
        }
        return speeds;
    }

    static void PrintSpeeds(Dictionary<string, (double read, double write)> speeds) {
        Console.WriteLine("\n" + new string('─', 60));
        Console.WriteLine($"\u001B[36mДиск   Чтение (МБ/с)   Запись (МБ/с)   Загрузка\u001B[0m");
        Console.WriteLine(new string('─', 60));

        foreach (var kv in speeds) {
            double read = kv.Value.read;
            double write = kv.Value.write;
            double load = Math.Min((read + write) / 1000 * 100, 100);

            string readColor = read > 300 ? "\u001B[32m" : read > 100 ? "\u001B[33m" : "\u001B[31m";
            string writeColor = write > 300 ? "\u001B[32m" : write > 100 ? "\u001B[33m" : "\u001B[31m";

            int barLen = 20;
            int filled = (int)(load / 100 * barLen);
            string bar = new string('█', filled) + new string('░', barLen - filled);

            Console.WriteLine($"{kv.Key,-6} {readColor}{read,8:F1}\u001B[0m     {writeColor}{write,8:F1}\u001B[0m     {bar} {load,5:F0}%");
        }
    }

    static void PrintStats() {
        Console.WriteLine($"\n\u001B[36m📊 Статистика:\u001B[0m");
        foreach (var kv in stats) {
            var s = kv.Value;
            if (s.ReadSpeeds.Count > 0) {
                double avgRead = s.ReadSpeeds.Average();
                double avgWrite = s.WriteSpeeds.Average();
                Console.WriteLine($"  {kv.Key}:");
                Console.WriteLine($"    Средняя скорость чтения: {avgRead:F1} МБ/с");
                Console.WriteLine($"    Средняя скорость записи: {avgWrite:F1} МБ/с");
                Console.WriteLine($"    Пиковая скорость чтения: {s.MaxRead:F1} МБ/с");
                Console.WriteLine($"    Пиковая скорость записи: {s.MaxWrite:F1} МБ/с");
            }
        }
    }
}

File: disk_speed.go
// disk_speed.go — Go версия

package main

import (
	"fmt"
	"os"
	"os/signal"
	"strconv"
	"syscall"
	"time"

	"github.com/shirou/gopsutil/v3/disk"
)

type DiskSpeedMonitor struct {
	interval time.Duration
	stats    map[string]*DiskStats
}

type DiskStats struct {
	ReadSpeeds  []float64
	WriteSpeeds []float64
	MaxRead     float64
	MaxWrite    float64
}

func NewDiskSpeedMonitor(interval time.Duration) *DiskSpeedMonitor {
	return &DiskSpeedMonitor{
		interval: interval,
		stats:    make(map[string]*DiskStats),
	}
}

func (m *DiskSpeedMonitor) getDiskIO() (map[string]disk.IOCountersStat, error) {
	io, err := disk.IOCounters()
	if err != nil {
		return nil, err
	}
	return io, nil
}

func (m *DiskSpeedMonitor) calculateSpeed(prev, curr map[string]disk.IOCountersStat, interval float64) map[string]map[string]float64 {
	speeds := make(map[string]map[string]float64)
	for diskName, currStats := range curr {
		if prevStats, ok := prev[diskName]; ok {
			readMB := float64(currStats.ReadBytes-prevStats.ReadBytes) / (1024 * 1024)
			writeMB := float64(currStats.WriteBytes-prevStats.WriteBytes) / (1024 * 1024)
			readSpeed := readMB / interval
			writeSpeed := writeMB / interval

			speeds[diskName] = map[string]float64{
				"read":  readSpeed,
				"write": writeSpeed,
			}

			if _, ok := m.stats[diskName]; !ok {
				m.stats[diskName] = &DiskStats{
					ReadSpeeds:  []float64{},
					WriteSpeeds: []float64{},
				}
			}
			m.stats[diskName].ReadSpeeds = append(m.stats[diskName].ReadSpeeds, readSpeed)
			m.stats[diskName].WriteSpeeds = append(m.stats[diskName].WriteSpeeds, writeSpeed)
			if readSpeed > m.stats[diskName].MaxRead {
				m.stats[diskName].MaxRead = readSpeed
			}
			if writeSpeed > m.stats[diskName].MaxWrite {
				m.stats[diskName].MaxWrite = writeSpeed
			}
		}
	}
	return speeds
}

func (m *DiskSpeedMonitor) printSpeeds(speeds map[string]map[string]float64) {
	fmt.Println("\n" + "─"*60)
	fmt.Printf("\x1b[36mДиск   Чтение (МБ/с)   Запись (МБ/с)   Загрузка\x1b[0m\n")
	fmt.Println("─" * 60)

	for disk, speed := range speeds {
		read := speed["read"]
		write := speed["write"]
		load := (read + write) / 1000 * 100
		if load > 100 {
			load = 100
		}

		readColor := "\x1b[32m"
		if read < 100 {
			readColor = "\x1b[31m"
		} else if read < 300 {
			readColor = "\x1b[33m"
		}

		writeColor := "\x1b[32m"
		if write < 100 {
			writeColor = "\x1b[31m"
		} else if write < 300 {
			writeColor = "\x1b[33m"
		}

		barLen := 20
		filled := int(load / 100 * float64(barLen))
		bar := ""
		for i := 0; i < filled; i++ {
			bar += "█"
		}
		for i := filled; i < barLen; i++ {
			bar += "░"
		}

		fmt.Printf("%-6s %s%8.1f\x1b[0m     %s%8.1f\x1b[0m     %s %5.0f%%\n",
			disk, readColor, read, writeColor, write, bar, load)
	}
}

func (m *DiskSpeedMonitor) printStats() {
	fmt.Printf("\n\x1b[36m📊 Статистика:\x1b[0m\n")
	for disk, stats := range m.stats {
		if len(stats.ReadSpeeds) > 0 {
			avgRead := 0.0
			avgWrite := 0.0
			for _, v := range stats.ReadSpeeds {
				avgRead += v
			}
			avgRead /= float64(len(stats.ReadSpeeds))
			for _, v := range stats.WriteSpeeds {
				avgWrite += v
			}
			avgWrite /= float64(len(stats.WriteSpeeds))

			fmt.Printf("  %s:\n", disk)
			fmt.Printf("    Средняя скорость чтения: %.1f МБ/с\n", avgRead)
			fmt.Printf("    Средняя скорость записи: %.1f МБ/с\n", avgWrite)
			fmt.Printf("    Пиковая скорость чтения: %.1f МБ/с\n", stats.MaxRead)
			fmt.Printf("    Пиковая скорость записи: %.1f МБ/с\n", stats.MaxWrite)
		}
	}
}

func (m *DiskSpeedMonitor) run() {
	fmt.Printf("\x1b[36m💾 Disk Speed Monitor (Go)\x1b[0m\n")
	fmt.Printf("Интервал: %.0f сек\n", m.interval.Seconds())
	fmt.Println("Нажмите Ctrl+C для остановки...")

	prev, err := m.getDiskIO()
	if err != nil {
		fmt.Printf("\x1b[31m❌ Ошибка: %v\x1b[0m\n", err)
		return
	}
	time.Sleep(m.interval)

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	ticker := time.NewTicker(m.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			curr, err := m.getDiskIO()
			if err != nil {
				fmt.Printf("\x1b[31m❌ Ошибка: %v\x1b[0m\n", err)
				continue
			}
			speeds := m.calculateSpeed(prev, curr, m.interval.Seconds())
			m.printSpeeds(speeds)
			prev = curr
		case <-sigChan:
			fmt.Println("\n\n⏹️ Остановка...")
			m.printStats()
			return
		}
	}
}

func main() {
	interval := 2
	if len(os.Args) > 1 {
		if val, err := strconv.Atoi(os.Args[1]); err == nil && val > 0 {
			interval = val
		}
	}
	monitor := NewDiskSpeedMonitor(time.Duration(interval) * time.Second)
	monitor.run()
}
